### DISCUSSION TITLE
[DataValidation] TimePicker does not trigger DataValidation on SelectedTime

### DISCUSSION THREAD

**Discussion by lamber-k** [AvaloniaUI/Avalonia#13172] [Q&A]
Hi,

First of, I'd like to thank the Avalonia team and the contributors for the amazing work you're doing. The experience so far has been great !

I'm struggling using DataValidation with TimePicker. This is working great with TextBox and NumericUpDown controls, but the TimePicker does not show any error, even if there is. I'm using ReactiveUI.Validation with `INotifyDataErrorInfo`.

On the ViewModel side, I've declared a Reactive Property, and a validation rule ion the constructor:

```
class MyViewModel : ReactiveValidationObject
{
    [Reactive]
    public TimeSpan? Duration { get; set; }

    public MyViewModel()
    {
         this.ValidationRule(x => x.Duration,
                duration => duration.HasValue,
                "My Error Message");
    }
}
```

On the view, I have a TimePicker bound to the Reactive Property:
```
<TimePicker SelectedTime="{Binding Duration}"/>
```

As a result, I don't see any error, even if the Duration does not have a value:
![image](https://github.com/AvaloniaUI/Avalonia/assets/3035083/a5b7528b-63a8-40ac-b1ad-54d9a375841a)

Looking at the DevTool, The `DataValidationErrors.Errors` Attached Property is `null`, and `DataValidationErrors.HasErrors` is `false`, even if the `DataContext.HasErrors` is `true` and the ValidationContext contains the validation rule error:

> Errors: My Error Message

The code I used to display this error is the following:
```
this.ValidationContext.Changed
    .Do(_ => Debug.WriteLine($"Errors: {ValidationContext.Text.ToSingleLine()}"))
    .Subscribe();
```

Time Picker DataContext - HasErrors property
![image](https://github.com/AvaloniaUI/Avalonia/assets/3035083/cc595c99-d83e-4f79-92d0-cdf88c8494da)

TimePicker control - DataValidationsErrors Attached Properties
![image](https://github.com/AvaloniaUI/Avalonia/assets/3035083/7d29b875-aec1-4613-9514-0068b28bdfe4)

I've take a look to the differences between TextBox/NumericUpDown controls vs. TimePicker, and I found that the `SelectedTimeProperty` does not `enableDataValidation` to true when registering.

NumericUpDown: https://github.com/AvaloniaUI/Avalonia/blob/11.0.4/src/Avalonia.Controls/NumericUpDown/NumericUpDown.cs#L109
TimePicker: https://github.com/AvaloniaUI/Avalonia/blob/11.0.4/src/Avalonia.Controls/DateTimePickers/TimePicker.cs#L45

Also, the NumericUpDown control override the `UpdateDataValidation` method to Set the error in `DataValidationErrors` when the Text or Value property has its value updated (https://github.com/AvaloniaUI/Avalonia/blob/11.0.4/src/Avalonia.Controls/NumericUpDown/NumericUpDown.cs#L406)

I'm not sure if this is a bug, or the `SelectedTimeProperty` is supposed to not trigger the DataValidation.
If so, how do we display the validation error ?
I've tried to create a custom control using the Documentation provided here: https://docs.avaloniaui.net/docs/next/guides/custom-controls/how-to-create-advanced-custom-controls#datavalidation-support but with no luck. I'm not sure how my "CustomTimePicker" control is supposed to be have all the TimePicker behaviors, but override the SelectedValueProperty data validation.

Thanks for your help,
Kevin

**Comment by Enigmatrix** [comment]
I think you should open an issue for this. In the meantime, I've had success for a similar problem using the solution below (`ComboBox`'s `SelectedValue` not having data validation)

```c#
public class ExtendedComboBox : ComboBox
{
    static ExtendedComboBox()
    {
        // enable data validation
        SelectedValueProperty.OverrideMetadata<ExtendedComboBox>(
            new StyledPropertyMetadata<object?>(defaultBindingMode: BindingMode.TwoWay, enableDataValidation: true));
    }

    // make sure we use the same style as a regular combobox
    protected override Type StyleKeyOverride { get; } = typeof(ComboBox);

    // not sure if this is necessary. previously, this was only SelectedItemProperty
    protected override void UpdateDataValidation(AvaloniaProperty property, BindingValueType state, Exception? error)
    {
        if (property == SelectedItemProperty || property == SelectedValueProperty)
            DataValidationErrors.SetError(this, error);
    }
}
```

**Comment by ILD3225** [comment]
Hello, let's enable validation through reflection until the developers include it in the source code)

```
public class TimePickerEx : TimePicker
{
    static TimePickerEx()
    {
        // Назначаем через рефлексию EnableDataValidation = true для SelectedTimeProperty, т.к. в исходном варианте не используется валидация
        var propertyMetadata = SelectedTimeProperty.GetMetadata(typeof(TimePickerEx));
        if (propertyMetadata.EnableDataValidation == false)
        {
            var declaringType = typeof(StyledPropertyMetadata<TimeSpan?>).GetProperty(nameof(AvaloniaPropertyMetadata.EnableDataValidation))?.DeclaringType;

            // <EnableDataValidation>k__BackingField
            const string fieldName = $"<{nameof(AvaloniaPropertyMetadata.EnableDataValidation)}>k__BackingField";
            var field = declaringType!.GetField(fieldName, BindingFlags.Instance | BindingFlags.NonPublic);
            field!.SetValue(propertyMetadata, true);
        }
    }

    protected override Type StyleKeyOverride => typeof(TimePicker);

    protected override void UpdateDataValidation(AvaloniaProperty property, BindingValueType state, Exception? error)
    {
        base.UpdateDataValidation(property, state, error);

        if (property == SelectedTimeProperty)
            DataValidationErrors.SetError(this, error);
    }
}
```

**Reply by timunie** [reply]
One could also file a PR instead of waiting and hack-fix a solution.

**Reply by ILD3225** [reply]
#14406
